cmd/encryption-v1.go

```diff
@@ -923,16 +924,18 @@ func getDecryptedETag(headers http.Header, objInfo ObjectInfo, copySource bool)
 	if crypto.IsMultiPart(objInfo.UserDefined) {
 		return objInfo.ETag
 	}
+
+	// If source is SSE-C encrypted and this is a SSE-C Copy operation
+	// we should return the last 32 bytes for the ETag, as there is no
+	// decryption necessary to comply with AWS S3 spec.
 	if crypto.SSECopy.IsRequested(headers) {
-		key, err = crypto.SSECopy.ParseHTTP(headers)
-		if err != nil {
-			return objInfo.ETag
-		}
+		return objInfo.ETag[len(objInfo.ETag)-32:]
```

Comment:
@vadmeste, while this will fix the precondition validation error, it will fail in the case of key rotation. If source and dest are the same - only the metadata is replaced and ETag is just copied over ( which is why it was being decrypted if copySource was true). In this case, ETag will be re-encrypted using the last 32 bytes rather than the md5sum, leading to incompatibility with s3 spec if ssec->s3 copy is done.

Reply:
We can detect let's say if the destination is being encrypted using SSE-S3 then we can try to get the key out of the request if not this code is fine.

Reply:
seems to me ObjectInfo needs to store encrypted md5 separately so that a helper function can get decrypted ETag if encryption key is passed in. The issue is ObjectInfo will only have decrypted ETag after GetObjectReader returns. We need both decrypted ETag (for key rotation) and last 32 bytes ( for pre-condition validation) in case of SSE-C

Reply:
yes, but that refactor is not needed if we simply add one check here..

Reply:
@poornas are you saying that ETag doesn't change when we rotate keys ? I tested with this script but it looks like the ETag is always changed:

```sh
#!/bin/bash

aws s3api put-object \
  --bucket vadmeste-garage --key my-secret-source-object \
  --sse-customer-algorithm AES256 \
  --sse-customer-key MzJieXRlc2xvbmdzZWNyZXRrZXltdXN0cHJvdmlkZWQ= \
  --sse-customer-key-md5 7PpPLAK26ONlVUGOWlusfg== \
  --body ./README.md

aws s3api head-object --bucket vadmeste-garage --key my-secret-source-object \
 --sse-customer-algorithm AES256 \
 --sse-customer-key MzJieXRlc2xvbmdzZWNyZXRrZXltdXN0cHJvdmlkZWQ= \
 --sse-customer-key-md5 7PpPLAK26ONlVUGOWlusfg==

aws s3api copy-object \
    --bucket vadmeste-garage --key my-secret-source-object \
    --copy-source /vadmeste-garage/my-secret-source-object \
    --metadata-directive REPLACE \
    --sse-customer-algorithm AES256 \
    --sse-customer-key "Qi1sHpQnppeJAo5WyM8w/BGJXJmTO/LBg6dcjThY6nM=" \
    --sse-customer-key-md5 "IcpiSupe46fQ0fb5AGQ2RQ==" \
    --copy-source-sse-customer-algorithm AES256 \
    --copy-source-sse-customer-key MzJieXRlc2xvbmdzZWNyZXRrZXltdXN0cHJvdmlkZWQ= \
    --copy-source-sse-customer-key-md5 7PpPLAK26ONlVUGOWlusfg==


aws s3api head-object --bucket vadmeste-garage --key my-secret-source-object \
 --sse-customer-algorithm AES256 \
 --sse-customer-key "Qi1sHpQnppeJAo5WyM8w/BGJXJmTO/LBg6dcjThY6nM=" \
 --sse-customer-key-md5 "IcpiSupe46fQ0fb5AGQ2RQ=="

```

Reply:
The reason is that the etag encrypted with a different key. So the encrypted etags will differ but the plain etag is the same.